src/stan/lang/ast/base_expr_type_def.hpp

```diff
@@ -0,0 +1,180 @@
+                                   base_type)
+      : base_type_(base_type) {
+    }
+
+    base_expr_type::base_expr_type(const vector_type&
+                                   base_type)
+      : base_type_(base_type) {
+    }
+
+    base_expr_type::base_expr_type(const row_vector_type&
+                                   base_type)
+      : base_type_(base_type) {
+    }
+
+    base_expr_type::base_expr_type(const matrix_type&
+                                   base_type)
+      : base_type_(base_type) {
+    }
+
+    bool base_expr_type::is_ill_formed_type() const {
+      is_ill_formed_type_vis v;
+      return boost::apply_visitor(v, base_type_);
+    }
+
+    bool base_expr_type::is_void_type() const {
+      is_void_type_vis v;
+      return boost::apply_visitor(v, base_type_);
+    }
+
+    bool base_expr_type::is_int_type() const {
+      is_int_type_vis v;
+      return boost::apply_visitor(v, base_type_);
+    }
+
+    bool base_expr_type::is_double_type() const {
+      is_double_type_vis v;
+      return boost::apply_visitor(v, base_type_);
+    }
+
+    bool base_expr_type::is_vector_type() const {
+      is_vector_type_vis v;
+      return boost::apply_visitor(v, base_type_);
+    }
+
+    bool base_expr_type::is_row_vector_type() const {
+      is_row_vector_type_vis v;
+      return boost::apply_visitor(v, base_type_);
+    }
+
+    bool base_expr_type::is_matrix_type() const {
+      is_matrix_type_vis v;
+      return boost::apply_visitor(v, base_type_);
+    }
+
+    bool base_expr_type::operator==(const base_expr_type& base_type) const {
+      if ((is_ill_formed_type() &&
+           base_type.is_ill_formed_type())
+          || (is_void_type() &&
+              base_type.is_void_type())
+          || (is_int_type() &&
+              base_type.is_int_type())
+          || (is_double_type() &&
+              base_type.is_double_type())
+          || (is_vector_type() &&
+              base_type.is_vector_type())
+          || (is_row_vector_type() &&
+              base_type.is_row_vector_type())
+          || (is_matrix_type() &&
+              base_type.is_matrix_type()))
+        return true;
+      return false;
+    }
+
+    bool base_expr_type::operator!=(const base_expr_type& base_type) const {
+      if ((is_ill_formed_type() &&
+           base_type.is_ill_formed_type())
+          || (is_void_type() &&
+              base_type.is_void_type())
+          || (is_int_type() &&
+              base_type.is_int_type())
+          || (is_double_type() &&
+              base_type.is_double_type())
+          || (is_vector_type() &&
+              base_type.is_vector_type())
+          || (is_row_vector_type() &&
+              base_type.is_row_vector_type())
+          || (is_matrix_type() &&
+              base_type.is_matrix_type()))
+        return false;
+      return true;
+    }
+
+    bool base_expr_type::operator<(const base_expr_type& base_type) const {
+      if (is_ill_formed_type()) {
```

Comment:
fixed.
added member var `order_id_`
set in templated constructor.
tried to make `order_id_` `const`, but this triggered a compiler warning because of the way these classes are used by the parser in `semantic_actions_def.cpp` and elsewhere.